mitosheet/src/mito/components/SearchBar.tsx

```diff
@@ -10,71 +10,225 @@ import Input from './elements/Input';
     setUIState: React.Dispatch<React.SetStateAction<UIState>>;
     uiState: UIState;
     mitoAPI: MitoAPI;
+    containerDiv: HTMLDivElement | null;
+    scrollAndRenderedContainerDiv: HTMLDivElement | null;
+    sheetView: SheetView;
+    gridState: GridState;
 }
 
 export const SearchBar: React.FC<SearchBarProps> = (props) => {
-    const { setUIState, uiState, mitoAPI } = props;
+    const {
+        setUIState,
+        uiState,
+        mitoAPI,
+        containerDiv,
+        scrollAndRenderedContainerDiv,
+        sheetView,
+        gridState
+    } = props;
+    const { searchValue, currentMatchIndex, matches } = uiState.currOpenSearch;
+
+    // totalMatches is undefined when we're making the API call to get the total number of matches.
     const [totalMatches, setTotalMatches] = React.useState<number | undefined>(undefined);
-    const searchValue = uiState.currOpenSearch.searchValue;
+    const [ showCautionMessage, setShowCautionMessage ] = React.useState<boolean>(false);
+
+    const scrollMatchIntoView = (match?: { row: number; col: number }) => {
+        // Columns have row index -1, so we check for that first.
+        if (match?.row === -1) {
+            scrollColumnIntoView(
+                containerDiv,
+                scrollAndRenderedContainerDiv,
+                sheetView,
+                gridState,
+                match?.col
+            )
+        } else if (match !== undefined) {
+            ensureCellVisible(
+                containerDiv,
+                scrollAndRenderedContainerDiv,
+                sheetView,
+                gridState,
+                match.row,
+                match.col,
+            )
+        }
+    }
 
-    /**
-     * Update the total number of matches when the search value changes
-     */
+    // Call the backend to get the new match information when the search value or sheet index changes.
     useDebouncedEffect(() => {
-        // If the search value is empty, set the total matches to 0
+        // If the search value is empty, set the total matches to 0 and don't call the API.
         if (searchValue === undefined || searchValue === '') {
             setTotalMatches(0);
             return;
         }
+
+        // Call the API to get the total number of matches and the displayed matches. 
         void mitoAPI.getTotalNumberMatches(uiState.selectedSheetIndex, searchValue ?? '').then((response) => {
             if ('error' in response) {
                 return;
             }
-            const total = response.result;
-            if (total !== undefined && !isNaN(Number(total))) {
-                setTotalMatches(Number(total));
+            const new_total_number_matches = response.result.total_number_matches;
+            const new_matches = response.result.matches;
+            // Update the total matches. 
+            if (!isNaN(Number(new_total_number_matches))) {
+                setTotalMatches(Number(new_total_number_matches));
+            }
```

Comment:
Why do you need to do this casting? I feel like if you get a number back, it should always be a number. I don't see this in the type system.

Reply:
Oh you're totally right, I assumed it'd be a string for some reason